### PR TITLE
[WIP] SLA and POC

### PR DESCRIPTION
**Pull request by kkellerlbl**
Please don't merge yet!

This doc is a first pass at formalizing a Service Level Agreement (SLA) and Point of Contact (POC) procedures for reporting and responding to problems on devops-supported systems.  Please leave comments here so we can make revisions to the doc before formally including it in project_guides.

### CODE REVIEW

**Comment by kkellerlbl** [comment]
One question I had is whether we would be able to ask the POC to marshal and perform non-emergency deployments in production.  My initial thought is that planned deployments should be done as part of a sprint, and therefore a devops person should be on that sprint team to help organize what needs to be deployed and when.  However, once that team gets to the deployment stage, if the POC is not busy putting out fires, he could be asked to help with the actual deploy.  If there were POC issues those would take priority over sprint deploys.

My worry would be that if two sprint teams are planning for production deploys there could be contention for the POC's free time.  I'm not sure how we'd want to resolve that.

**Comment by kkellerlbl** [comment]
It'd also be nice to have a calendar, maybe we can do a Google calendar.  But that's not quite as important, especially if we pin in #poc who's the POC.

**Comment by nlharris** [comment]
Keith, this document looks reasonable to me.

**Comment by fperez** [comment]
This looks good to me, @kkellerlbl.  I'd suggest you post to the kbase-all list so others can comment on this PR, so we can have a timeline for agreement/merge...

**Comment by kkellerlbl** [comment]
I think Shane and Dan wanted to discuss the SLA a little more before trying to solicit comments from the wider group.  But yes, once that's done we should post to all before attempting to merge.

**Comment by fperez** [comment]
OK, great. I'd like a general policy of having any PR on this repo being given say ~3-5 business days after a post to kbase-all for review and feedback before a merge is attempted. It could be the low end of that for simple things, and the higher end for more potentially controversial things.

Since this is effectively imposing a hard contract on people's time commitments, I'd give it 5 business days of open discussion before a merge _after_ you ping the list, so there's ample time for anyone to comment.

**Comment by fperez** [comment]
BTW, ping @dangunter and @scanon here since you're interested in their feedback :)

**Comment by kkellerlbl** [comment]
@danielolson5 not Dan G.  :)

**Comment by danielolson5** [comment]
We should be able to circulate a final version next week - please hold for now.  

Dan

**Comment by kkellerlbl** [comment]
Would it make more sense to have changes that I'm waiting for input for before announcing on all to just be in a fork?  This way a PR isn't hanging over the repo that's still waiting for a small group discussion before releasing to the wider group for comment.

**Comment by fperez** [comment]
I don't think it's a problem to have the PR open, you did mark it as not
ready for merge yet... We do that all the time in IPython, it lets us use
the same PR mechanism for everything.

**Comment by fperez** [comment]
I edited the title to [WIP], short for 'work in progress', so when we see the list of PRs we don't bother clicking on it if we're looking for things to merge, only if we want to review/comment.

Once it's ready to merge, we can remove that marker.

**Comment by scanon** [comment]
Dan: Are there any major changes you would like to make?

**Comment by danielolson5** [comment]
I proposed some changes on kkellers fork.  

pagerduty needs to be reworded to fit what we can setup with the nersc monitoring center.  

Dan

**Comment by kkellerlbl** [comment]
Dan, is that this commit on your fork?

https://github.com/danielolson5/project_guides/commit/dfdca1d100a8b550ce809f073d63c6762d7ddb45

**Comment by danielolson5** [comment]
Yes, thats it.  

Dan

**Comment by fperez** [comment]
@kkellerlbl, could you review/merge Dan's commits into your branch? That way, we can include them into this PR directly and make them part of the discussion...

**Comment by kkellerlbl** [comment]
I just merged Dan's changes.

I would like to try to keep some sort of reference to Slack in there somehow, but I know Dan is less excited about using Slack for the POC role than I am.  OTOH I'm fine pulling this in now and revisiting the Slack question later.

**Comment by fperez** [comment]
Up to you, folks, this should reflect your sense as a team :)

**Comment by kkellerlbl** [comment]
Okay, thanks Fernando.  @danielolson5 if you're okay with it you can merge yourself or ask me to do.

File: SLA_and_POC.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Service Level Agreement (SLA) for KBase
+
+Our normal operating hours are 8am to 7pm Eastern time (7am-6pm Central, 6am-5pm Pacific), Monday through Friday (except holidays).  During these hours, the POC will respond to (not necessarily resolve) production issues within 2 hours.  Outside of these hours, the POC will respond as soon as possible the next business day.  For catastrophic issues that occur outside our normal SLA, the POC will respond as soon as possible.  (Catastrophic issues are those which cause most or all of KBase to be completely unusable with no possible workaround.  Examples include a core service such as workspace or Shock going down.)  Production issues will be routed by the help desk coordinator, internal developers, and the monitoring center to the pager email address.
+
+Issues in non-production environments can be reported to help@kbase.us.  The helpdesk coordinator will route a Jira ticket to the devops component.  These issues will be routed to the most appropriate resource by the POC within one business day.
+
+The POC rotation will be managed in pagerduty.
```
Comment (kkellerlbl) [inline SLA_and_POC.md]: In GCal instead?

File: SLA_and_POC.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Service Level Agreement (SLA) for KBase
+
+Our normal operating hours are 8am to 7pm Eastern time (7am-6pm Central, 6am-5pm Pacific), Monday through Friday (except holidays).  During these hours, the POC will respond to (not necessarily resolve) production issues within 2 hours.  Outside of these hours, the POC will respond as soon as possible the next business day.  For catastrophic issues that occur outside our normal SLA, the POC will respond as soon as possible.  (Catastrophic issues are those which cause most or all of KBase to be completely unusable with no possible workaround.  Examples include a core service such as workspace or Shock going down.)  Production issues will be routed by the help desk coordinator, internal developers, and the monitoring center to the pager email address.
+
+Issues in non-production environments can be reported to help@kbase.us.  The helpdesk coordinator will route a Jira ticket to the devops component.  These issues will be routed to the most appropriate resource by the POC within one business day.
```
Comment (kkellerlbl) [inline SLA_and_POC.md]: Use JIRA, not help@kbase

File: SLA_and_POC.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+The POC rotation will be managed in pagerduty. 
+
+The POC will not be assigned a sprint team or other duties during his or her rotation.  The primary duty of the POC will be to respond to reported issues, refer them to other devops team members as needed, and follow up with the reporter to verify a resolution.  Secondary duties may include doing nonintrusive maintenance on KBase systems.
```
Comment (kkellerlbl) [inline SLA_and_POC.md]: We probably need some more flexibility here, like on the sprint team but secondary to POC duties.

File: SLA_and_POC.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Service Level Agreement (SLA) for KBase
+
+Our normal operating hours are 8am to 7pm Eastern time (7am-6pm Central, 6am-5pm Pacific), Monday through Friday (except holidays).  During these hours, the POC will respond to (not necessarily resolve) production issues within 2 hours.  Outside of these hours, the POC will respond as soon as possible the next business day.  For catastrophic issues that occur outside our normal SLA, the POC will respond as soon as possible.  (Catastrophic issues are those which cause most or all of KBase to be completely unusable with no possible workaround.  Examples include a core service such as workspace or Shock going down.)  Production issues will be routed by the help desk coordinator, internal developers, and the monitoring center to the pager email address.
```
Comment (kkellerlbl) [inline SLA_and_POC.md]: Hours are of course negotiable!

File: SLA_and_POC.md (hunk 2 above)
Comment (kkellerlbl) [inline SLA_and_POC.md]: Probably also want to have a #poc channel in Slack

Comment (nlharris) [reply SLA_and_POC.md]: YES

**Pull request merged**